src/twisted/web/http.py

```diff
@@ -1985,23 +1996,38 @@ def _dataReceived_CRLF(self) -> bool:
 
     def _dataReceived_TRAILER(self) -> bool:
         """
-        Await the carriage return and line feed characters that follow the
-        terminal zero-length chunk. Then invoke C{finishCallback} and switch to
-        state C{'FINISHED'}.
+        Collect trailer headers if received and finish at the terminal zero-length
+        chunk. Then invoke C{finishCallback} and switch to state C{'FINISHED'}.
 
         @returns: C{False}, as there is either insufficient data to continue,
             or no data remains.
-
-        @raises _MalformedChunkedDataError: when anything other than CRLF is
-            received.
         """
-        if len(self._buffer) < 2:
+        if (
+            self._receivedTrailerHeadersSize + len(self._buffer)
+            > self._totalTrailerHeadersSize
+        ):
+            raise _MalformedChunkedDataError("Trailer headers data is too long")
+
+        eolIndex = self._buffer.find(b"\r\n", self._start)
+
+        if eolIndex == -1:
+            # Restart the search upon receipt of more data at the start of the
+            # new data, minus one in case the last character of the buffer is
+            # CR.
+            self._start = len(self._buffer) - 1
             return False
 
-        if not self._buffer.startswith(b"\r\n"):
-            raise _MalformedChunkedDataError("Chunk did not end with CRLF")
+        if eolIndex > 0:
+            self._trailerHeaders.append(self._buffer[0:eolIndex])
```

Comment:
The current implementation is ok.

Maybe we can just check for size after this append.
In this way, we don't need the extra `self._receivedTrailerHeadersSize`

Reply:
In case if server response with infinite trailer header without CRLF then memory will be filled completely. So eolIndex will always equal to -1.

Reply:
Thanks.

Can we have an autoamted test for this case?

It looks like the current automated tests is not covering this  case.

The current implementation might be ok, but the future, someone might refactor the code and the existing tests will still pass, while a bug is introduced.

------

I will have to check the code.
I still don't know why we can't just check the size of `self._buffer` and why we need an extra variable...
but maybe there is a good reason.

Note that the goal of raisin an exception is not to be matematically correct, but only to prevent a memory attack.

Reply:
I added new unit test for unfinished trailer headers.

About current logic: in this extra variable we accumulate number of consumed bytes. When we meet CRLF characters we shrink buffer and increase this extra variable. How we would know how many characters we consumed without extra variable? Buffer self._buffer shrinks on every trailer header (every met CRLF chars)

Reply:
So the extra variable solves 2 problems:
1. Monitor number of chars which was sent before CRLF. And if there is too many such characters algorithm throws exception.
2. If there is a lot of trailer headers the extra variable accumulates the lengths of this headers. `self._buffer` shrinks on every trailer header so we can't use `self._buffer` for detection of length of received data.